commandsv3/src/main/java/org/wpilib/commands3/Scheduler.java

```diff
@@ -0,0 +1,891 @@
+  }
+
+  /**
+   * Schedules a command to run. If a running command schedules another command (for example,
+   * parallel groups will do this), then the new command is assumed to be a bound child of the
+   * running command. Child commands will automatically be cancelled by the scheduler when their
+   * parent command completes or is canceled. Child commands will also immediately begin execution,
+   * without needing to wait for the next {@link #run()} invocation. This allows highly nested
+   * compositions to begin running the actual meaningful commands sooner without needing to wait one
+   * scheduler cycle per nesting level; with the default 20ms update period, 5 levels of nesting
+   * would be enough to delay actions by 100 milliseconds - instead of only 20.
+   *
+   * <p>Does nothing if the command is already scheduled or running, or requires at least one
+   * mechanism already used by a higher priority command.
+   *
+   * <p>If one command schedules another ("parent" and "fork"), the forked command will be canceled
+   * when the parent command completes. It is not possible to fork a command and have it live longer
+   * than the command that forked it.
+   *
+   * @param command the command to schedule
+   * @return the result of the scheduling attempt. See {@link ScheduleResult} for details.
+   * @throws IllegalArgumentException if scheduled by a command composition that has already
+   *     scheduled another command that shares at least one required mechanism
+   */
+  public ScheduleResult schedule(Command command) {
+    // Note: we use a throwable here instead of Thread.currentThread().getStackTrace() for easier
+    //       stack frame filtering and modification.
+    var binding =
+        new Binding(
+            BindingScope.global(), BindingType.IMMEDIATE, command, new Throwable().getStackTrace());
+
+    return schedule(binding);
+  }
+
+  // Package-private for use by Trigger
+  ScheduleResult schedule(Binding binding) {
+    var command = binding.command();
+
+    if (isScheduledOrRunning(command)) {
+      return ScheduleResult.ALREADY_RUNNING;
+    }
+
+    if (!isSchedulable(command)) {
+      return ScheduleResult.LOWER_PRIORITY_THAN_RUNNING_COMMAND;
+    }
+
+    for (var scheduledState : m_onDeck) {
+      if (!command.conflictsWith(scheduledState.command())) {
+        // No shared requirements, skip
+        continue;
+      }
+      if (command.isLowerPriorityThan(scheduledState.command())) {
+        // Lower priority than an already-scheduled (but not yet running) command that requires at
+        // one of the same mechanism. Ignore it.
+        return ScheduleResult.LOWER_PRIORITY_THAN_RUNNING_COMMAND;
+      }
+    }
+
+    // Evict conflicting on-deck commands
+    // We check above if the input command is lower priority than any of these,
+    // so at this point we're guaranteed to be >= priority than anything already on deck
+    evictConflictingOnDeckCommands(command);
```

Comment:
Double pass could be optimized away by collecting commands in an else block of `if (isLowerPriorityThan)` and then iterating over that.